rewatch/src/project_context.rs

```diff
@@ -0,0 +1,288 @@
+use crate::build::packages;
+use crate::config::Config;
+use crate::helpers;
+use ahash::AHashSet;
+use anyhow::Result;
+use anyhow::anyhow;
+use log::debug;
+use std::fmt;
+use std::path::{Path, PathBuf};
+
+type RescriptJsonPath = PathBuf;
+
+/// Represents the context of a command, categorizing in various ways:
+/// - SingleProject: one rescript.json with no parent or children.
+/// - MonorepoRoot: a rescript.json with (dev-)dependencies found in subfolders.
+/// - MonorepoPackage: a rescript.json with a parent rescript.json.
+///   The current package name is listed in the parent (dev-)dependencies.
+// I don't think this linting rule matters very much as we only create a single instance of this enum.
+// See https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
+#[allow(clippy::large_enum_variant)]
+pub enum ProjectContext {
+    /// Single project - no workspace relationship
+    SingleProject { config: Config, path: RescriptJsonPath },
+    /// Monorepo root - contains local dependencies (symlinked in node_modules)
+    MonorepoRoot {
+        config: Config,
+        path: RescriptJsonPath,
+        local_dependencies: AHashSet<String>, // names of local deps
+        local_dev_dependencies: AHashSet<String>,
+    },
+    /// Package within a monorepo - has a parent workspace
+    MonorepoPackage {
+        config: Config,
+        // This field is only used for debugging purposes, maybe it could be removed in the future.
+        path: RescriptJsonPath,
+        parent_config: Config,
+        parent_path: RescriptJsonPath,
+    },
+}
+
+fn format_dependencies(dependencies: &AHashSet<String>) -> String {
+    if dependencies.is_empty() {
+        String::from("[]")
+    } else {
+        format!(
+            "[\n{}\n]",
+            dependencies
+                .iter()
+                .map(|dep| format!("  \"{dep}\""))
+                .collect::<Vec<String>>()
+                .join(",\n")
```

Comment:
`format` + `iter` + `join` can be done more efficiently with `write!` without any additional intermediate overhead;

```
let mut output = String::from("[\n{");
for dep of dependencies {
  fmt::write(&mut output, format!("  \"{dep}\,\n"")).expect("Error");
}
fmt::write(&mut output, "}\n]").expect("Error");

output
```

If you really want to fly - I think multiple `write` operations in the loop would be even faster than serializing the `dep` with formatter, as it's already a string - but I'm not quite sure how much faster it would be.

Reply:
Ah - I see it's used within `fmt` only - perhaps push it into that context, then we'll know it's not used for anything outside of formatting / performance matters less